PatternPal/PatternPal/Services/StepByStepService.cs

```diff
@@ -1,41 +1,211 @@
+
+        foreach (string document in request.Documents)
+        {
+            graph.AddFile(
+                FileManager.MakeStringFromFile(document),
+                document);
+        }
+
+        graph.CreateGraph();
+
+        // Add not only the name of class but also the filepath
+        GetSelectableClassesResponse selectableClasses = new();
+        foreach ((string source, _) in graph.GetAll().OrderByDescending(p => File.GetLastWriteTime(p.Value.GetRoot().GetSource())))
+        {
+            selectableClasses.SelectableClasses.Add(source);
+        }
+
+        return Task.FromResult(selectableClasses);
     }
 
+    /// <inheritdoc />
     public override Task< GetInstructionByIdResponse > GetInstructionById(
         GetInstructionByIdRequest request,
         ServerCallContext context)
     {
-        throw new NotImplementedException();
+        IInstruction instructionSet = InstructionSetHolder.instructionSet.Steps[ (int)request.InstructionNumber ];
+        GetInstructionByIdResponse response = new()
+                                              {
+                                                  Instruction = new Instruction
+                                                                {
+                                                                    Title = instructionSet.Requirement,
+                                                                    Description = instructionSet.Description,
+                                                                    FileId = ""
+                                                                }
+                                              };
+
+        return Task.FromResult(response);
     }
 
+    /// <inheritdoc />
     public override Task< CheckInstructionResponse > CheckInstruction(
         CheckInstructionRequest request,
         ServerCallContext context)
+    {
+        // Make the graph based on the users opened solution or files.
+        SyntaxGraph graph = new();
+        foreach (string document in request.Documents)
+        {
+            graph.AddFile(
+                FileManager.MakeStringFromFile(document),
+                document);
+        }
+        graph.CreateGraph();
+
+        Dictionary<string, IEntity> allEntities = graph.GetAll();
+        // Selected item in the combobox file path obtained from the graph.
+        string selectedNodePath = string.Empty;
+        if(request.SelectedItem != "")
+        {
+            selectedNodePath = allEntities[request.SelectedItem].GetRoot().GetSource();
+        }
+        // Otherwise the one in viewModel.
+        else
+        {
+            selectedNodePath = request.Documents[0];
+        }
+        
+        RecognizerRunner runner = new(
+            selectedNodePath,
+            InstructionSetHolder.instructionSet.Steps[request.InstructionId]);
+        IList<ICheckResult> res = runner.Run(pruneAll: true);
+        NodeCheckResult checkRes = (NodeCheckResult)res[0];
```

Comment:
This crashes if the list is empty.

Reply:
I now check whether any of the provided files has a correct implementation, if the list was empty (aka no roots provided to the runner) then the implementation was also correct.